validphys2/src/validphys/photon/compute.py

```diff
@@ -70,132 +86,131 @@ def __init__(self, theoryid, lux_params, replicas):
+
+          photon_qin = np.array(
+            [lux.EvaluatePhoton(x, self.q_in**2).total for x in XGRID]
         )
-        photon_qin += self.generate_errors(replica)
-        # fiatlux computes x * gamma(x)
-        photon_qin /= XGRID
-        # TODO : the different x points could be even computed in parallel
-
-        # Load eko and reshape it
-        with EKO.read(self.path_to_eko_photon) as eko_photon:
-            # TODO : if the eko has not the correct grid we have to reshape it
-            # it has to be done inside vp-setupfit
-
-            # NB: the eko should contain a single operator
-            for _, elem in eko_photon.items():
-                eko_op = elem.operator
-
-                pdfs_init = np.zeros_like(eko_op[0, 0])
-                for j, pid in enumerate(basis_rotation.flavor_basis_pids):
-                    if pid == 22:
-                        pdfs_init[j] = photon_qin
-                        ph_id = j
-                    elif pid not in self.luxpdfset.flavors:
-                        continue
-                    else:
-                        pdfs_init[j] = np.array(
-                            [self.luxpdfset.xfxQ(x, self.q_in, photonreplica, pid) / x for x in XGRID]
-                        )
-
-                pdfs_final = np.einsum("ajbk,bk", eko_op, pdfs_init)
-
-        photon_Q0 = pdfs_final[ph_id]
-
-        # we want x * gamma(x)
-        return XGRID * photon_Q0
+          photon_qin += self.generate_errors(replica)
+
+          # fiatlux computes x * gamma(x)
+          photon_qin /= XGRID
+
+          # Load eko and reshape it
+          with EKO.read(self.path_to_eko_photon) as eko_photon:
+              # TODO : if the eko has not the correct grid we have to reshape it
+              # it has to be done inside vp-setupfit
+              
+              # NB: the eko should contain a single operator
+              for _, elem in eko_photon.items():
+                  eko_op = elem.operator
+
+                  pdfs_init = np.zeros_like(eko_op[0, 0])
+                  for j, pid in enumerate(basis_rotation.flavor_basis_pids):
+                      if pid == 22:
+                          pdfs_init[j] = photon_qin
+                          ph_id = j
+                      elif pid not in self.luxpdfset.flavors:
+                          continue
+                      else:
+                          pdfs_init[j] = np.array(
+                              [self.luxpdfset.xfxQ(x, self.q_in, photonreplica, pid) / x for x in XGRID]
+                          )
+
+                  pdfs_final = np.einsum("ajbk,bk", eko_op, pdfs_init)
+
+          photon_Q0 = pdfs_final[ph_id]
+          photon_array = XGRID * photon_Q0
+
+          if self.save_to_disk:
+              np.savez_compressed(path_to_photon / f"replica_{photonreplica}.npz",photon_array=photon_array)
+              log.info(f"Saved photon replica {photonreplica} to {path_to_photon}")
```

Comment:
A cheap way of solving the "one-eko-per-thread" problem here might be simply doing the save_to_disk early (saving the photon from `pdfs_init` instead of from `pdfs_final` and then doing the eko during n3fit.

(so, if you have `self.save_to_disk` then don't do the eko part, if you don't, do the eko)